docs/proposal/2022-12-23-rbac.md

```diff
@@ -0,0 +1,399 @@
+
+- User identification:
+  - User email
+  - Static value for a given configuration
+- Group identification:
+  - Channel name
+  - User groups (Slack, Mattermost), Roles (Discord). Not supported by MS Teams.
+  - Static value for a given configuration
+
+We can support just a few options initially, and later add more based on user feedback.
+
+To get user email, channel name or user group name, we would need to do additional API calls in bot logic. However, we can cache the fetched values.
+
+#### Implementation details
+
+Slack:
+
+- We need to add two new scopes to the Slack app: `users:read` and `users:read.email`
+- We need to do additional API calls to get email (user info for a given ID) and channel name (conversation info for a given ID)
+- Unfortunately Slack user groups are in paid plan, so I couldn't test them. Probably we will need one additional call to get user group details to get its name.
+
+Mattermost:
+
+- We need to do additional API call to get user email based on its ID
+- [Based on user ID we can get the groups](https://api.mattermost.com/#tag/groups/operation/GetGroupsByUserId). They are not included in the message data.
+
+Discord:
+
+- We need to query user data as they are missing from the message object. Quoting the docs: "The field user won't be included in the member object attached to MESSAGE_CREATE and MESSAGE_UPDATE gateway events."
+- [For each message we get role IDs](https://discord.com/developers/docs/topics/gateway-events#message-create).
+
+  ```go
+  &discordgo.MessageCreate{
+    Message: &discordgo.Message{
+      // ...
+      Member: &discordgo.Member{
+        // ...
+        Roles: []string{
+          "976789858670497805", // <-- this is what we can use
+        },
+      },
+    },
+  }
+  ```
+
+  Based on that we can query for a full role object to get its name.
+
+  ```go
+  b.api.AddHandler(func(s *discordgo.Session, m *discordgo.MessageCreate) {
+      msg := discordMessage{
+        Event: m,
+      }
+      roles, _ := b.api.GuildRoles(m.GuildID)
+      // manually filter role by ID - unfortunately there's just list capability - couldn't find get
+          // we can cache the roles as they are server-wide (or guild-wide, depends what term is used)
+      // get role.Name - more: https://discord.com/developers/docs/topics/permissions#role-object
+  ```
+
+MS Teams:
+
+- We need to do additional API calls to get the user name email. Firstly, we need to [retrieve user ID and/or channel ID](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/channel-and-group-conversations?tabs=json#retrieve-mentions), then [get user details](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/get-teams-context?tabs=json#get-single-member-details).
+- There are no user groups or roles in MS Teams we can use.
+
+### Configure mapping
+
+User creates Roles and ClusterRoles, as well as the RoleBindings and ClusterRoleBindings. For example:
```

Comment:
This means, Botkube Helm will be creating that automatically right?

Reply:
For the default configuration yes, we will enable static mapping by default:

From consequences:

>  - Update defaults to use static mapping.
>  - Enable creating optional ClusterRole for all users to keep previous behavior.

Users (Cluster admins) will be able to create custom (cluster)role + (cluster)rolebindings per channel and that would be up to the users, unless the configuration is managed and then we can facilitate it further.